File: vscode/tests/unit/copilotQuota.test.ts
/**
 * copilotQuota.test.ts
 *
 * Unit tests for src/copilotQuota.ts — covers the pure response parsing
 * (all four fallback shapes), HTTP status mapping, session orchestration,
 * and the in-memory cache. Uses the built-in Node.js test runner
 * (node:test + node:assert), same as gameEngine.test.ts.
 */

import { describe, it, beforeEach } from "node:test";
import assert from "node:assert/strict";
import {
  parseQuotaBody,
  getCachedCopilotQuota,
  clearCopilotQuotaCache,
  COPILOT_QUOTA_CACHE_TTL_MS,
  type FetchLike,
  type GetSessionLike,
} from "../../src/copilotQuota";

/** Build a fake `fetch` that returns a fixed status + JSON body, recording calls. */
function fakeFetch(status: number, body: unknown, calls: string[]): FetchLike {
  return (async (url: string) => {
    calls.push(String(url));
    return {
      status,
      ok: status >= 200 && status < 300,
      json: async () => body,
    } as unknown as Response;
  }) as FetchLike;
}

/** Build a fake `fetch` that throws (simulates an offline/network failure). */
function throwingFetch(): FetchLike {
  return (async () => {
    throw new Error("network down");
  }) as FetchLike;
}

/** Build a fake `fetch` that returns 200 with an unparseable body. */
function unparseableFetch(): FetchLike {
  return (async () => {
    return {
      status: 200,
      ok: true,
      json: async () => { throw new Error("bad json"); },
    } as unknown as Response;
  }) as FetchLike;
}

/** Build a fake session getter that returns a fixed token (or undefined). */
function fakeGetSession(token: string | undefined, calls: boolean[]): GetSessionLike {
  return (async (createIfNone: boolean) => {
    calls.push(createIfNone);
    return token !== undefined ? { accessToken: token } : undefined;
  }) as GetSessionLike;
}

describe("parseQuotaBody", () => {
  it("parses the modern quota_snapshots shape", () => {
    const result = parseQuotaBody({
      quota_snapshots: { premium_interactions: { percent_remaining: 42.6, unlimited: false } },
    });
    assert.deepEqual(result, { ok: true, percentRemaining: 43, unlimited: false });
  });

  it("parses the modern shape with unlimited: true", () => {
    const result = parseQuotaBody({
      quota_snapshots: { premium_interactions: { percent_remaining: 100, unlimited: true } },
    });
    assert.deepEqual(result, { ok: true, percentRemaining: 100, unlimited: true });
  });

  it("parses the legacy limited_user_quotas.premium_interactions shape", () => {
    const result = parseQuotaBody({
      limited_user_quotas: { premium_interactions: { limit: 300, used: 75 } },
    });
    assert.deepEqual(result, { ok: true, percentRemaining: 75, unlimited: false });
  });

  it("parses the legacy limited_user_quotas.completions shape with explicit remaining", () => {
    const result = parseQuotaBody({
      limited_user_quotas: { completions: { monthly_maximum: 300, used: 100, remaining: 150 } },
    });
    assert.deepEqual(result, { ok: true, percentRemaining: 50, unlimited: false });
  });

  it("parses the nested quota shape", () => {
    const result = parseQuotaBody({ quota: { maximum: 100, used: 20 } });
    assert.deepEqual(result, { ok: true, percentRemaining: 80, unlimited: false });
  });

  it("parses the nested premium_interactions shape", () => {
    const result = parseQuotaBody({ premium_interactions: { total: 50, used: 10 } });
    assert.deepEqual(result, { ok: true, percentRemaining: 80, unlimited: false });
  });

  it("parses the flat premium_requests_maximum shape", () => {
    const result = parseQuotaBody({ premium_requests_maximum: 200, premium_requests_used: 50 });
    assert.deepEqual(result, { ok: true, percentRemaining: 75, unlimited: false });
  });

  it("parses the flat monthly_maximum_premium_requests shape", () => {
    const result = parseQuotaBody({ monthly_maximum_premium_requests: 10, premium_requests_used: 10 });
    assert.deepEqual(result, { ok: true, percentRemaining: 0, unlimited: false });
  });

  it("returns 0% (not a divide-by-zero crash) for a zero-limit legacy quota", () => {
    const result = parseQuotaBody({ limited_user_quotas: { premium_interactions: { limit: 0, used: 0 } } });
    assert.deepEqual(result, { ok: true, percentRemaining: 0, unlimited: false });
  });

  it("returns parse_error for an unrecognized body shape", () => {
    assert.deepEqual(parseQuotaBody({ some: "unrelated field" }), { ok: false, reason: "parse_error" });
  });

  it("returns parse_error for a non-object body", () => {
    assert.deepEqual(parseQuotaBody("nope"), { ok: false, reason: "parse_error" });
    assert.deepEqual(parseQuotaBody(null), { ok: false, reason: "parse_error" });
  });
});

describe("getCachedCopilotQuota", () => {
  beforeEach(() => {
    clearCopilotQuotaCache();
  });

  it("returns no_session when getSession resolves undefined", async () => {
    const calls: boolean[] = [];
    const result = await getCachedCopilotQuota(fakeGetSession(undefined, calls), fakeFetch(200, {}, []), true);
    assert.deepEqual(result, { ok: false, reason: "no_session" });
    assert.deepEqual(calls, [true]);
  });

  it("passes createIfNone through to the session getter", async () => {
    const calls: boolean[] = [];
    await getCachedCopilotQuota(fakeGetSession(undefined, calls), fakeFetch(200, {}, []), false);
    assert.deepEqual(calls, [false]);
  });

  it("fetches and parses on success", async () => {
    const fetchCalls: string[] = [];
    const body = { quota_snapshots: { premium_interactions: { percent_remaining: 60, unlimited: false } } };
    const result = await getCachedCopilotQuota(
      fakeGetSession("tok_abc", []),
      fakeFetch(200, body, fetchCalls),
      true
    );
    assert.deepEqual(result, { ok: true, percentRemaining: 60, unlimited: false });
    assert.equal(fetchCalls.length, 1);
    assert.match(fetchCalls[0], /copilot_internal\/user/);
  });

  it("maps 401 to unauthorized", async () => {
    const result = await getCachedCopilotQuota(fakeGetSession("tok", []), fakeFetch(401, {}, []), true);
    assert.deepEqual(result, { ok: false, reason: "unauthorized" });
  });

  it("maps 403 to unauthorized", async () => {
    const result = await getCachedCopilotQuota(fakeGetSession("tok", []), fakeFetch(403, {}, []), true);
    assert.deepEqual(result, { ok: false, reason: "unauthorized" });
  });

  it("maps other non-2xx statuses to network_error", async () => {
    const result = await getCachedCopilotQuota(fakeGetSession("tok", []), fakeFetch(500, {}, []), true);
    assert.deepEqual(result, { ok: false, reason: "network_error" });
  });

  it("maps a thrown fetch to network_error", async () => {
    const result = await getCachedCopilotQuota(fakeGetSession("tok", []), throwingFetch(), true);
    assert.deepEqual(result, { ok: false, reason: "network_error" });
  });

  it("maps an unparseable JSON body to parse_error", async () => {
    const result = await getCachedCopilotQuota(fakeGetSession("tok", []), unparseableFetch(), true);
    assert.deepEqual(result, { ok: false, reason: "parse_error" });
  });

  it("caches a successful outcome and does not re-fetch within the TTL", async () => {
    const sessionCalls: boolean[] = [];
    const fetchCalls: string[] = [];
    const body = { quota_snapshots: { premium_interactions: { percent_remaining: 60, unlimited: false } } };
    const getSession = fakeGetSession("tok", sessionCalls);
    const fetchImpl = fakeFetch(200, body, fetchCalls);

    const first = await getCachedCopilotQuota(getSession, fetchImpl, true, 1_000);
    const second = await getCachedCopilotQuota(getSession, fetchImpl, true, 1_000 + COPILOT_QUOTA_CACHE_TTL_MS - 1);

    assert.deepEqual(first, second);
    assert.equal(sessionCalls.length, 1);
    assert.equal(fetchCalls.length, 1);
  });

  it("re-fetches once the TTL has elapsed", async () => {
    const sessionCalls: boolean[] = [];
    const fetchCalls: string[] = [];
    const body = { quota_snapshots: { premium_interactions: { percent_remaining: 60, unlimited: false } } };
    const getSession = fakeGetSession("tok", sessionCalls);
    const fetchImpl = fakeFetch(200, body, fetchCalls);

    await getCachedCopilotQuota(getSession, fetchImpl, true, 1_000);
    await getCachedCopilotQuota(getSession, fetchImpl, true, 1_000 + COPILOT_QUOTA_CACHE_TTL_MS + 1);

    assert.equal(sessionCalls.length, 2);
    assert.equal(fetchCalls.length, 2);
  });

  it("does not cache a transient network_error, so the next call retries", async () => {
    const result1 = await getCachedCopilotQuota(fakeGetSession("tok", []), throwingFetch(), true, 1_000);
    assert.deepEqual(result1, { ok: false, reason: "network_error" });

    const fetchCalls: string[] = [];
    const body = { quota_snapshots: { premium_interactions: { percent_remaining: 60, unlimited: false } } };
    const result2 = await getCachedCopilotQuota(fakeGetSession("tok", []), fakeFetch(200, body, fetchCalls), true, 1_001);
    assert.deepEqual(result2, { ok: true, percentRemaining: 60, unlimited: false });
    assert.equal(fetchCalls.length, 1);
  });
});

File: pycharm/src/test/kotlin/com/codotchi/CopilotQuotaTest.kt
package com.codotchi

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

/**
 * Unit tests for CopilotQuota.kt — mirrors the coverage of
 * vscode/tests/unit/copilotQuota.test.ts. Only the pure parsing logic and the
 * injectable-RawHttp orchestration/cache/device-flow state machine are
 * tested — real network I/O is never exercised.
 */
class CopilotQuotaTest {

    @AfterEach
    fun clearCache() {
        CopilotQuotaCache.clear()
    }

    // ── parseQuotaBody ───────────────────────────────────────────────────────

    @Test
    fun `parseQuotaBody parses the modern quota_snapshots shape`() {
        val body = mapOf("quota_snapshots" to mapOf("premium_interactions" to mapOf("percent_remaining" to 42.6, "unlimited" to false)))
        assertEquals(CopilotQuotaResult.Ok(43, false), parseQuotaBody(body))
    }

    @Test
    fun `parseQuotaBody parses the modern shape with unlimited true`() {
        val body = mapOf("quota_snapshots" to mapOf("premium_interactions" to mapOf("percent_remaining" to 100.0, "unlimited" to true)))
        assertEquals(CopilotQuotaResult.Ok(100, true), parseQuotaBody(body))
    }

    @Test
    fun `parseQuotaBody parses the legacy premium_interactions shape`() {
        val body = mapOf("limited_user_quotas" to mapOf("premium_interactions" to mapOf("limit" to 300.0, "used" to 75.0)))
        assertEquals(CopilotQuotaResult.Ok(75, false), parseQuotaBody(body))
    }

    @Test
    fun `parseQuotaBody parses the legacy completions shape with explicit remaining`() {
        val body = mapOf("limited_user_quotas" to mapOf("completions" to mapOf("monthly_maximum" to 300.0, "used" to 100.0, "remaining" to 150.0)))
        assertEquals(CopilotQuotaResult.Ok(50, false), parseQuotaBody(body))
    }

    @Test
    fun `parseQuotaBody parses the nested quota shape`() {
        val body = mapOf("quota" to mapOf("maximum" to 100.0, "used" to 20.0))
        assertEquals(CopilotQuotaResult.Ok(80, false), parseQuotaBody(body))
    }

    @Test
    fun `parseQuotaBody parses the nested premium_interactions shape`() {
        val body = mapOf("premium_interactions" to mapOf("total" to 50.0, "used" to 10.0))
        assertEquals(CopilotQuotaResult.Ok(80, false), parseQuotaBody(body))
    }

    @Test
    fun `parseQuotaBody parses the flat premium_requests_maximum shape`() {
        val body = mapOf("premium_requests_maximum" to 200.0, "premium_requests_used" to 50.0)
        assertEquals(CopilotQuotaResult.Ok(75, false), parseQuotaBody(body))
    }

    @Test
    fun `parseQuotaBody parses the flat monthly_maximum_premium_requests shape`() {
        val body = mapOf("monthly_maximum_premium_requests" to 10.0, "premium_requests_used" to 10.0)
        assertEquals(CopilotQuotaResult.Ok(0, false), parseQuotaBody(body))
    }

    @Test
    fun `parseQuotaBody returns 0 percent for a zero-limit legacy quota instead of dividing by zero`() {
        val body = mapOf("limited_user_quotas" to mapOf("premium_interactions" to mapOf("limit" to 0.0, "used" to 0.0)))
        assertEquals(CopilotQuotaResult.Ok(0, false), parseQuotaBody(body))
    }

    @Test
    fun `parseQuotaBody returns ParseError for an unrecognized body shape`() {
        assertEquals(CopilotQuotaResult.ParseError("Unrecognized Copilot quota response shape"), parseQuotaBody(mapOf("some" to "unrelated field")))
    }

    // ── fetchQuota (injected RawHttp) ────────────────────────────────────────

    @Test
    fun `fetchQuota parses a successful response`() {
        val http = RawHttp { _, _, _, _ -> 200 to """{"quota_snapshots":{"premium_interactions":{"percent_remaining":60,"unlimited":false}}}""" }
        assertEquals(CopilotQuotaResult.Ok(60, false), fetchQuota("tok", http))
    }

    @Test
    fun `fetchQuota maps 401 and 403 to Unauthorized`() {
        val http401 = RawHttp { _, _, _, _ -> 401 to "" }
        val http403 = RawHttp { _, _, _, _ -> 403 to "" }
        assertEquals(CopilotQuotaResult.Unauthorized, fetchQuota("tok", http401))
        assertEquals(CopilotQuotaResult.Unauthorized, fetchQuota("tok", http403))
    }

    @Test
    fun `fetchQuota maps other non-2xx statuses to NetworkError`() {
        val http = RawHttp { _, _, _, _ -> 500 to "" }
        assertEquals(CopilotQuotaResult.NetworkError("HTTP 500"), fetchQuota("tok", http))
    }

    @Test
    fun `fetchQuota maps a thrown RawHttp to NetworkError`() {
        val http = RawHttp { _, _, _, _ -> throw RuntimeException("network down") }
        assertEquals(CopilotQuotaResult.NetworkError("network down"), fetchQuota("tok", http))
    }

    @Test
    fun `fetchQuota maps an unparseable body to ParseError`() {
        val http = RawHttp { _, _, _, _ -> 200 to "not json" }
        assertTrue(fetchQuota("tok", http) is CopilotQuotaResult.ParseError)
    }

    // ── CopilotQuotaCache ────────────────────────────────────────────────────

    @Test
    fun `CopilotQuotaCache does not re-fetch within the TTL`() {
        var calls = 0
        val fetch = { calls++; CopilotQuotaResult.Ok(60, false) as CopilotQuotaResult }
        CopilotQuotaCache.get(1_000L, fetch)
        CopilotQuotaCache.get(1_000L + COPILOT_QUOTA_CACHE_TTL_MS - 1, fetch)
        assertEquals(1, calls)
    }

    @Test
    fun `CopilotQuotaCache re-fetches once the TTL has elapsed`() {
        var calls = 0
        val fetch = { calls++; CopilotQuotaResult.Ok(60, false) as CopilotQuotaResult }
        CopilotQuotaCache.get(1_000L, fetch)
        CopilotQuotaCache.get(1_000L + COPILOT_QUOTA_CACHE_TTL_MS + 1, fetch)
        assertEquals(2, calls)
    }

    @Test
    fun `CopilotQuotaCache does not cache a transient NetworkError`() {
        var calls = 0
        val fetch = { calls++; CopilotQuotaResult.NetworkError("boom") as CopilotQuotaResult }
        CopilotQuotaCache.get(1_000L, fetch)
        CopilotQuotaCache.get(1_001L, fetch)
        assertEquals(2, calls)
    }

    // ── Device flow ──────────────────────────────────────────────────────────

    @Test
    fun `parseDeviceCodeResponse reads all fields with defaults for missing optionals`() {
        val body = mapOf("device_code" to "dc", "user_code" to "ABCD-1234", "verification_uri" to "https://github.com/login/device")
        val result = parseDeviceCodeResponse(body)
        assertEquals(DeviceCodeResponse("dc", "ABCD-1234", "https://github.com/login/device", 900, 5), result)
    }

    @Test
    fun `parseDeviceCodeResponse falls back to verification_uri_complete when verification_uri is absent`() {
        val body = mapOf("device_code" to "dc", "user_code" to "ABCD-1234", "verification_uri_complete" to "https://github.com/login/device?x=1", "expires_in" to 600.0, "interval" to 10.0)
        val result = parseDeviceCodeResponse(body)
        assertEquals(DeviceCodeResponse("dc", "ABCD-1234", "https://github.com/login/device?x=1", 600, 10), result)
    }

    @Test
    fun `pollForToken returns Success once access_token appears`() {
        var call = 0
        val http = RawHttp { _, _, _, _ ->
            call++
            if (call < 3) 200 to """{"error":"authorization_pending"}"""
            else 200 to """{"access_token":"gho_abc123"}"""
        }
        val result = CopilotDeviceFlow.pollForToken(
            deviceCode = "dc", intervalSeconds = 1, expiresInSeconds = 60,
            http = http, sleep = { /* no-op in tests */ }, nowMs = { 0L },
        )
        assertEquals(TokenPollResult.Success("gho_abc123"), result)
        assertEquals(3, call)
    }

    @Test
    fun `pollForToken returns AccessDenied when the user declines`() {
        val http = RawHttp { _, _, _, _ -> 200 to """{"error":"access_denied"}""" }
        val result = CopilotDeviceFlow.pollForToken("dc", 1, 60, http, sleep = {}, nowMs = { 0L })
        assertEquals(TokenPollResult.AccessDenied, result)
    }

    @Test
    fun `pollForToken returns Expired when the code expires server-side`() {
        val http = RawHttp { _, _, _, _ -> 200 to """{"error":"expired_token"}""" }
        val result = CopilotDeviceFlow.pollForToken("dc", 1, 60, http, sleep = {}, nowMs = { 0L })
        assertEquals(TokenPollResult.Expired, result)
    }

    @Test
    fun `pollForToken returns Expired when the deadline passes without a terminal response`() {
        val http = RawHttp { _, _, _, _ -> 200 to """{"error":"authorization_pending"}""" }
        var now = 0L
        val result = CopilotDeviceFlow.pollForToken(
            deviceCode = "dc", intervalSeconds = 5, expiresInSeconds = 10,
            http = http, sleep = { now += it }, nowMs = { now },
        )
        assertEquals(TokenPollResult.Expired, result)
    }

    @Test
    fun `pollForToken returns Error on an unrecognized error code`() {
        val http = RawHttp { _, _, _, _ -> 200 to """{"error":"something_else","error_description":"weird"}""" }
        val result = CopilotDeviceFlow.pollForToken("dc", 1, 60, http, sleep = {}, nowMs = { 0L })
        assertEquals(TokenPollResult.Error("weird"), result)
    }

    @Test
    fun `pollForToken returns Error when RawHttp throws`() {
        val http = RawHttp { _, _, _, _ -> throw RuntimeException("network down") }
        val result = CopilotDeviceFlow.pollForToken("dc", 1, 60, http, sleep = {}, nowMs = { 0L })
        assertEquals(TokenPollResult.Error("network down"), result)
    }
}
